src/java/org/apache/cassandra/service/ReplicaFilteringProtection.java

```diff
@@ -212,83 +156,122 @@ private UnfilteredPartitionIterator executeReadCommand(ReadCommand cmd, InetAddr
     {
-        return (partitionKey, versions) -> {
+        return new UnfilteredPartitionIterators.MergeListener()
+        {
+            @Override
+            public void close()
+            {
+                // If we hit the failure threshold before consuming a single partition, record the current rows cached.
+                tableMetrics.rfpRowsCachedPerQuery.update(Math.max(currentRowsCached, maxRowsCached));
+            }
 
-            PartitionBuilder[] builders = new PartitionBuilder[sources.length];
+            public UnfilteredRowIterators.MergeListener getRowMergeListener(DecoratedKey partitionKey, List<UnfilteredRowIterator> versions)
+            {
+                PartitionBuilder[] builders = new PartitionBuilder[sources.length];
 
-            for (int i = 0; i < sources.length; i++)
-                builders[i] = new PartitionBuilder(partitionKey, columns(versions), stats(versions));
+                for (int i = 0; i < sources.length; i++)
+                    builders[i] = new PartitionBuilder(partitionKey, sources[i], columns(versions), EncodingStats.merge(versions, NULL_TO_NO_STATS));
 
-            return new UnfilteredRowIterators.MergeListener()
-            {
-                @Override
-                public void onMergedPartitionLevelDeletion(DeletionTime mergedDeletion, DeletionTime[] versions)
+                return new UnfilteredRowIterators.MergeListener()
                 {
-                    // cache the deletion time versions to be able to regenerate the original row iterator
-                    for (int i = 0; i < versions.length; i++)
-                        builders[i].setDeletionTime(versions[i]);
-                }
+                    @Override
+                    public void onMergedPartitionLevelDeletion(DeletionTime mergedDeletion, DeletionTime[] versions)
+                    {
+                        // cache the deletion time versions to be able to regenerate the original row iterator
+                        for (int i = 0; i < versions.length; i++)
+                            builders[i].setDeletionTime(versions[i]);
+                    }
 
-                @Override
-                public Row onMergedRows(Row merged, Row[] versions)
-                {
-                    // cache the row versions to be able to regenerate the original row iterator
-                    for (int i = 0; i < versions.length; i++)
-                        builders[i].addRow(versions[i]);
+                    @Override
+                    public Row onMergedRows(Row merged, Row[] versions)
+                    {
+                        // cache the row versions to be able to regenerate the original row iterator
+                        for (int i = 0; i < versions.length; i++)
+                        {
+                            builders[i].addRow(versions[i]);
+                            currentRowsCached++;
+                            checkCachedRowThresholds();
```

Comment:
@maedhroz Currently the listener increases the cached rows count while the builder decreases it. It's also the builder how adds and removes things from the cache. Also, the listener accesses directly to `partitionRowsCached` while the builder  does it through a method. We could easily move these two lines to `RFP.PartitionBuilder#addRow` and so make the row merge listener free from all cache management stuff. Also we could have a `RFP#increaseCachedRows()` method combining these two lines, which would be the counterpart of `RFP#releaseCachedRows(int)`. You can see [here](https://github.com/adelapena/cassandra/commit/38780dda5721839333cdcb79b5c0a54e2d0fc845) how it would look like. Feel free to ignore it if you don't like it.